scripts/ont_calc_volumes.py

```diff
@@ -3,88 +3,176 @@
+                log.append(
+                    f"Basing calculations on 'Volume to take (uL)': {round(udf_tools.fetch(art_out, 'Volume to take (uL)'),2)}"
+                )
+                vol_to_take = min(udf_tools.fetch(art_out, "Volume to take (uL)"), vol)
+            else:
+                raise AssertionError(f"No target metrics specified for {art_out.name}")
 
+            # Based on volume to take, calculate corresponding amounts
             if conc_units == "nM":
-                amt_taken_fmol = vol_to_take * art_tuple[0]["uri"].udf["Concentration"]
-                amt_taken_ng = fmol_to_ng(amt_taken_fmol, size)
-
+                amt_taken_fmol = conc * vol_to_take
+                amt_taken_ng = formula.fmol_to_ng(amt_taken_fmol, size_bp)
             elif conc_units == "ng/ul":
-                amt_taken_ng = vol_to_take * art_tuple[0]["uri"].udf["Concentration"]
-                amt_taken_fmol = ng_to_fmol(amt_taken_ng, size)
+                amt_taken_ng = conc * vol_to_take
+                amt_taken_fmol = formula.ng_to_fmol(amt_taken_ng, size_bp)
+
+            log.append(f"--> 'Volume to take (uL)': {round(vol_to_take, 2)}")
+            log.append(f"--> 'Amount (ng)': {round(amt_taken_ng, 2)}")
+            log.append(f"--> 'Amount (fmol)': {round(amt_taken_fmol, 2)}")
+
+            # Populate fields
+            if "ONT Start Sequencing" in currentStep.type.name:
+                udf_tools.put(
+                    art_out,
+                    "ONT flow cell loading amount (fmol)",
+                    round(amt_taken_fmol, 2),
+                )
+                udf_tools.put(art_out, "Volume to take (uL)", round(vol_to_take, 2))
+            else:
+                udf_tools.put(art_out, "Amount (fmol)", round(amt_taken_fmol, 2))
+                udf_tools.put(art_out, "Amount (ng)", round(amt_taken_ng, 2))
+                udf_tools.put(art_out, "Volume to take (uL)", round(vol_to_take, 2))
+
+            log.append("\n")
+
+        # Write log
+        timestamp = dt.now().strftime("%y%m%d_%H%M%S")
+        log_filename = (
+            "_".join(["ont_volume_calc_log", currentStep.id, timestamp]) + ".txt"
+        )
+        with open(log_filename, "w") as logContext:
+            logContext.write("\n".join(log))
+
+        # Upload log
+        for out in currentStep.all_outputs():
+            if out.name == "Volume Calculation Log":
+                for f in out.files:
+                    lims.request_session.delete(f.uri)
+                lims.upload_new_file(out, log_filename)
 
-            if "Amount (ng)" in [t[0] for t in art_tuple[1]["uri"].udf.items()]:
-                art_tuple[1]["uri"].udf["Amount (ng)"] = round(amt_taken_ng,1)
-            art_tuple[1]["uri"].udf["Amount (fmol)"] = round(amt_taken_fmol,1)
-            art_tuple[1]["uri"].udf["Volume to take (uL)"] = round(vol_to_take,1)
-            art_tuple[1]["uri"].put()
+    except BaseException as e:
```

Comment:
I would probably move this try except outside the `main` 

```python
try:
    main(lims, args)
except BaseException as e:
    sys.stderr.write(str(e))
    sys.exit(2)

Reply:
[a7fbe27](https://github.com/SciLifeLab/scilifelab_epps/pull/250/commits/a7fbe278ee1f705687931e63c9d5d7f8717e9f0c)